Allow to optionally defer JS from custom layout templates

michalkleiner opened the pull request:
### Description:

Since login page doesn't use nearly any of the bundled JS, we can defer its loading and the form will still be visible quite early. Then with the JS already loaded continuing to the dashboard should use the cached files already.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Allow to optionally defer JS from custom layout templates

michalkleiner committed:
Use DOMContentLoaded event in iframeBuster

michalkleiner committed:
Use DOMContentLoaded event in piwikHelper

michalkleiner commented:
Relates to https://github.com/matomo-org/matomo/issues/20460

michalkleiner commented:
Replaces https://github.com/matomo-org/matomo/pull/20750 due to a source branch name change.

michalkleiner committed:
Remove posting of Matomo.afterInitialVueEntryProcess event

michalkleiner committed:
Tweak mechanism to wait for jQuery to be available

michalkleiner committed:
Remove superfluous anonymous function

michalkleiner committed:
Remove waiting for jQuery as it should be ready when DOMContentLoaded is triggered

michalkleiner committed:
Remove jQuery dependency from iframe buster

michalkleiner commented on plugins/CoreHome/templates/_uiControl.twig:
```diff
@@ -6,4 +6,8 @@
-<script>$(document).ready(function () { require('{{ jsNamespace }}').{{ jsClass }}.initElements(); });</script>
\ No newline at end of file
+<script>
+    document.addEventListener('DOMContentLoaded', function () {
+        $(document).ready(function () { require('{{ jsNamespace }}').{{ jsClass }}.initElements(); });
+    }, false);
```
Verified locally and via https://jsfiddle.net/9xnh71uy/ that we can have `$(document).ready` inside `DOMContentLoaded` event handler and it works as expected.

michalkleiner commented on plugins/Morpheus/templates/_iframeBuster.twig:
After discussion with Stefan this has been refactored to use vanilla JS and since it's included from inside <body> then we assume it exists and can be directly worked with.

The logic is also inverted to do nothing if Matomo is the top frame already.

sgiehl commented on plugins/Morpheus/templates/_iframeBuster.twig:
```diff
@@ -1,11 +1,10 @@
-        } else { top.location = self.location; }
-    });
+        if (self !== top) {
+            const theBody = document.getElementsByTagName('body')[0];
```
Hm. Not sure if we should use `const` here. I guess it should be valid for most modern browsers that might be used to view Matomo, but using `var` would be more safe though...

michalkleiner replied:
I checked https://caniuse.com/const vs https://caniuse.com/mdn-javascript_statements_var and based on the percentages decided it was safe to use.
Surely we can go back to `var`, though with that approach we would need to change it in more places, instead of a mix of `var`, `let` and `const`, so I think we can be a tad progressive here.

sgiehl replied:
In the past we actually used `var` for all cases that don't run through a compiler. It's safe to use that in all vue code, as it is replaced with `var` when it is compiled.

michalkleiner replied:
Switched back to `var`.

michalkleiner commented on plugins/Morpheus/templates/_iframeBuster.twig:
```diff
@@ -1,11 +1,10 @@
-        } else { top.location = self.location; }
-    });
+        if (self !== top) {
+            const theBody = document.getElementsByTagName('body')[0];
```
```suggestion
            var theBody = document.getElementsByTagName('body')[0];
```

michalkleiner committed:
Use `var` instead of `const` in iframe buster

sgiehl reviewed (changes requested):
Did some last testing and discovered one more issue we need to solve for this.
After the changes the `cachebuster` parameter is no longer appended to the defered loaded scripts.
The code to append it doesn't work anymore due to the added `defer`. Guess we should add the option defer to the replacements here:
https://github.com/matomo-org/matomo/blob/aa2674a8241d5e4e55a2f7fbb74292e06eab8780/core/View.php#L365-L380

michalkleiner committed:
Add replace pattern to keep defer attribute while supporting cache busting param

sgiehl commented on core/View.php:
```diff
@@ -364,6 +364,7 @@ protected function applyFilterCacheBuster($output)
 
         $pattern = array(
             '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"]+)[\'"]>~',
+            '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"]+)[\'"] defer>~',
```
This order actually breaks the chunked files as it already replaces what should be replaced by the rule that follows 2 lines below. Guess you need to put this line down the list.

michalkleiner replied:
Good spot @sgiehl, thanks for pointing that out. Of course we need to cater for the more specific first. Updated.

michalkleiner committed:
Move less specific defer line after more specific chunked files defer line

sgiehl commented on core/View.php:
```diff
@@ -366,6 +366,7 @@ protected function applyFilterCacheBuster($output)
             '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"]+)[\'"]>~',
             '~<script src=[\'"]([^\'"]+)[\'"] type=[\'"]text/javascript[\'"]>~',
             '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"]+?chunk=[^\'"]+)[\'"] defer>~',
+            '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"]+)[\'"] defer>~',
```
This now might append it twice to the chunked assets. Guess this should fix it:
```suggestion
            '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"?]*\?[^\'"]+)[\'"] defer>~',
            '~<script type=[\'"]text/javascript[\'"] src=[\'"]([^\'"?]+)[\'"] defer>~',
```

sgiehl commented on core/AssetManager.php:
```diff
@@ -138,20 +138,21 @@ public function getCssInclusionDirective()
             $this->getMergedNonCoreJSAsset()->delete();
 
-            $result .= $this->getIndividualCoreAndNonCoreJsIncludes();
+            $result .= $this->getIndividualCoreAndNonCoreJsIncludes($deferJS);
```
Guess we also need to set `deferJS` to `false` here. When merged assets are disabled we can't server the javascript assets defered. Problem is that they might get loaded in incorrect order that way. At least I'm seeing javascript errors having merged assets disabled.

michalkleiner replied:
I guess we can remove it here for now but if the order matters in how they are merged or how they are served unmerged, we probably should look into the whole system. But that the Matomo 6 thing.

Ok, I'll remove it here. Not just set to false, we can remove the whole mechanism here and in the nested calls as that's the only place where it's used.

sgiehl replied:
Our system ensure that certain scripts are loaded first, like jQuery or vue.js. When assets are merged they are placed first in the files. But when using unmerged assets another assets might get loaded faster, running into problems as jquery or vue might not be loaded yet.
As using unmerged assets is actually only used for development there shouldn't be any need to change anything around that.

michalkleiner committed:
Remove defer option when assets are not merged as not relevant there

Using the defer attribute when assets are not merged can cause order-related issues.

michalkleiner committed:
Correctly handle URLs with query params using defer

Co-authored-by: Stefan Giehl <stefan@matomo.org>

sgiehl reviewed (approved):
Seems to work as expected for me now 🎉
